Add node types

mcharanrm opened the pull request:
#### Highlights of the PR
1. Add `node_type` to controlplane and execution plane nodes.
2. Added few comment lines to `vars.yaml` to describe the default and optional `node_type` to a node.

jainnikhil30 commented on templates/inventory_cluster_4.1.j2:
```diff
@@ -12,7 +12,7 @@
 [execution_nodes]
 {% for execution_node in groups['execution'] %}
-{{ execution_node }}
+{{ execution_node }} node_type={{ execution_plane_node_type | default('execution') }}
```
With this all the nodes in execution_nodes group would either be execution or hop node?
But with a hop node we need atleast 1 execution node in the end right ?

mcharanrm replied:
yeah totally correct. let me fix it

mcharanrm replied:
a new commit https://github.com/redhat-performance/ibm_cloud_tower_deploy/pull/18/commits/34094f1ea69ee8a4101208652f895bf5f36a8b3b has been pushed to fix the either or situation. Now I have added flexibility to support different `node_type` to different instances in the inventory

mcharanrm commented:
I have create few additional variables, `tower_control_nodes`, `tower_hybrid_nodes`, `tower_execution_nodes`, and `tower_hop_nodes`which allows the flexibility to specify how many how many tower nodes would you like to have `node_type` as `control`and how many as `hybrid`. How many of execution nodes would you like to have `node_type` as `execution` and how many as `hop`. 

If you don't specify anything, then all nodes under `[automationcontroller]` will have the default `node_type=hybrid` and all nodes under `[execution_nodes]` will have the default `node_type=execution`

I have tested the changes using a dummy playbook and a dummy inventory and the results are clean.

1. with additional variables, `tower_control_nodes`, `tower_hybrid_nodes`, `tower_execution_nodes`, and `tower_hop_nodes`
```
[cmusali@cmusali ibm_cloud_tower_deploy]$ ansible-playbook -i inventory dummy_play.yaml -e tower_control_nodes=1 -e tower_hyrid_nodes=2 -e tower_execution_nodes=2 -e tower_hop_nodes=1

PLAY [localhost] *******************************************************************************************************************************************************************

TASK [template] ********************************************************************************************************************************************************************
changed: [localhost]

PLAY RECAP *************************************************************************************************************************************************************************
localhost                  : ok=1    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   

[cmusali@cmusali ibm_cloud_tower_deploy]$ cat dummy_inventory 
[automationcontroller]
tower_perf_test_1 node_type=control

tower_perf_test_2 node_type=hybrid
tower_perf_test_3 node_type=hybrid


[database]
database_perf_test_1
database_perf_test_2
database_perf_test_3

[execution_nodes]
execution_perf_test_1 node_type=execution
execution_perf_test_2 node_type=execution

execution_perf_test_3 node_type=hop


[execution_nodes:vars]
peers=automationcontroller
```

2. With defaults

```
[cmusali@cmusali ibm_cloud_tower_deploy]$ ansible-playbook -i inventory dummy_play.yaml 

PLAY [localhost] *******************************************************************************************************************************************************************

TASK [template] ********************************************************************************************************************************************************************
changed: [localhost]

PLAY RECAP *************************************************************************************************************************************************************************
localhost                  : ok=1    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   

[cmusali@cmusali ibm_cloud_tower_deploy]$ cat dummy_inventory 
[automationcontroller]
tower_perf_test_1 node_type=hybrid
tower_perf_test_2 node_type=hybrid
tower_perf_test_3 node_type=hybrid


[database]
database_perf_test_1
database_perf_test_2
database_perf_test_3

[execution_nodes]
execution_perf_test_1 node_type=execution
execution_perf_test_2 node_type=execution
execution_perf_test_3 node_type=execution


[execution_nodes:vars]
peers=automationcontroller

```

jainnikhil30 commented:
@mcharanrm 

Why are there 3 DB nodes ? 

```
[database]
database_perf_test_1
database_perf_test_2
database_perf_test_3
```
That doesnt look correct to me.

Also, we have another issue. So if you have a hop node, you would need to add a host_vars with peer/peers so that hop node knows which is the execution node/hop node it needs to connect to. I don't see that we are doing that. TBH, I would actually not add hop/execution node type at all. With hop node network topology can be very complex and its not possible for us to do that in an automated fashion. I would recommend we only allow control and hybrid node type, that way we can control what kind of control plane we want. Any node under execution_nodes will be an execution type node (isolated node). With this you will be able to make a comparison of 3.8 with isolated nodes (1 tower nodes 1 isolated node) and 4.1 (1 controlplane node type "control" and 1 execution node type "execution"). Please let me know if this makes sense ?

mcharanrm commented:
Ah I replicated the same testing scenario with dummy inventory template, dummy inventory to make sure the idea of adding `node_type` to targets doesn't create any mess during tower installation. 

My main focus in this PR is to show what it looks like adding node_type to control plane nodes and execution plane node but not database 

```
ansible-playbook -i inventory dummy_play.yaml  -e tower_control_nodes=1 \
                                               -e tower_hyrid_nodes=2 \
                                               -e tower_execution_nodes=2 \
                                               -e tower_hop_nodes=1
```

```
[automationcontroller]
tower_perf_test_1 node_type=control
tower_perf_test_2 node_type=hybrid
tower_perf_test_3 node_type=hybrid

[execution_nodes]
execution_perf_test_1 node_type=execution
execution_perf_test_2 node_type=execution
execution_perf_test_3 node_type=hop

[execution_nodes:vars]
peers=automationcontroller
```

>So if you have a hop node, you would need to add a host_vars with peer/peers so that hop node knows which is the execution node/hop node it needs to connect to. I don't see that we are doing that. TBH, I would actually not add hop/execution node type at all. With hop node network topology can be very complex and its not possible for us to do that in an automated fashion.

I have added `node_type=hop` assuming we may need it in the future but not for the immediate use since it require few other inventory changes. If we don't want hop nodes, we can go with defaults values and we can have all nodes in execution plane as execution nodes and all nodes in control plane as hybrid nodes.

```
ansible-playbook -i inventory dummy_play.yaml  
```
```
[automationcontroller]
tower_perf_test_1 node_type=hybrid
tower_perf_test_2 node_type=hybrid
tower_perf_test_3 node_type=hybrid

[execution_nodes]
execution_perf_test_1 node_type=execution
execution_perf_test_2 node_type=execution
execution_perf_test_3 node_type=execution

```

jainnikhil30 commented:
Oh okay. My bad I though that was an actual Tower inventory.
Okay cool in that case this looks good. But we need to make sure we create an issue on the repo saying we have the ability to set the node_type as hop but we dont have the bits and pieces that it needs in the inventory yet.

jainnikhil30 commented:
Thanks Charan :+1:
